File: 2024/06/part1.py
count = 0
seen = set()


def main():
    global grid
    with open("input.txt", "r") as file:
        grid = file.read().splitlines()

    i, j = findStart()
    move(i, j, 0, 0)
    print(len(seen))


dirs = [[-1, 0], [0, 1], [1, 0], [0, -1]]


def move(i: int, j: int, dir: int, depth: int):
    # Recursion doesn't work because python LUL
    while True:
        seen.add((i, j))
        a, b = dirs[dir]
        newi = i + a
        newj = j + b
        if newi < 0 or newi > len(grid) - 1:
            break
        if newj < 0 or newj > len(grid[i]) - 1:
            break

        nextSquare = grid[newi][newj]
        if nextSquare == "#":
            dir = (dir + 1) % len(dirs)
            depth += 1
        else:
            i = newi
            j = newj
            depth += 1


def findStart():
    i = 0
    for line in grid:
        j = line.find("^")
        if j != -1:
            return [i, j]
        i += 1

    return [-1, -1]


main()

File: 2024/06/ocaml/part1.ml
type grid = string array
type cell = int * int

module CellSet = Set.Make (struct
  type t = cell

  let compare (i, j) (h, k) =
    match (i, j, h, k) with
    | i, j, h, k when i == h -> j - k
    | i, j, h, k -> i - h
end)

let findStart gd =
  let ibound = Array.length gd in
  let jbound = String.length gd.(0) in

  let rec find i j =
    match (i, j) with
    | i, j when i == ibound -> failwith "bad find"
    | i, j when j == jbound -> find (i + 1) 0
    | i, j when gd.(i).[j] == '^' -> (i, j)
    | i, j -> find i (j + 1)
  in

  find 0 0

let isInBounds ibound jbound cell =
  let i, j = cell in
  (not (i < 0)) && (not (j < 0)) && i < ibound && j < jbound

let walkAndCount gd start =
  let dirs = [| (-1, 0); (0, 1); (1, 0); (0, -1) |] in
  let nextDir i = (i + 1) mod 4 in

  let ibound = Array.length gd in
  let jbound = String.length gd.(0) in

  let rec walk seen dir current =
    let di, dj = dirs.(dir) in
    let ci, cj = current in

    let next = (ci + di, cj + dj) in
    let newSeen = CellSet.add current seen in

    if isInBounds ibound jbound next then
      let ni, nj = next in
      if gd.(ni).[nj] == '#' then walk newSeen (nextDir dir) current
      else walk newSeen dir next
    else CellSet.cardinal newSeen
  in

  walk CellSet.empty 0 start

let () =
  let default = "input.txt" in
  let file = if Array.length Sys.argv == 2 then Sys.argv.(1) else default in
  let ic = open_in file in

  let s = really_input_string ic (in_channel_length ic) in
  let lines = s |> String.trim |> String.split_on_char '\n' in
  let gd = Array.of_list lines in

  let count = gd |> findStart |> walkAndCount gd in

  count |> Printf.printf "%d\n";

  close_in ic
